libcfnet/policy_server.c

```diff
@@ -187,3 +187,156 @@ const char *PolicyServerGetPort()
+}
+
+/**
+ * @brief     Reads the policy_server.dat file.
+ * @param[in] workdir the directory of policy_server.dat usually GetWorkDir()
+ * @return    Trimmed contents of policy_server.dat file. Null terminated. 
+ */
+char *PolicyServerReadFile(const char *workdir)
+{
+    char contents[CF_MAX_SERVER_LEN] = "";
+
+    char *filename = PolicyServerFilename(workdir);
+    FILE *fp = fopen(filename, "r");
+    if (fp == NULL)
+    {
+        Log( LOG_LEVEL_VERBOSE, "Could not open file '%s' (fopen: %s)",
+             filename, GetErrorStr() );
+        free(filename);
+        return NULL;
+    }
+
+    if (fgets(contents, CF_MAX_SERVER_LEN, fp) == NULL)
+    {
+        Log( LOG_LEVEL_VERBOSE, "Could not read file '%s' (fgets: %s)",
+             filename, GetErrorStr() );
+        free(filename);
+        fclose(fp);
+        return NULL;
+    }
+    fclose(fp);
+    char *start = TrimWhitespace(contents);
+    return xstrdup(start);
+}
+
+
+/**
+ * @brief      Reads and parses the policy_server.dat file.
+ *
+ * @code{.c}
+ * //Typical usage:
+ * char *host, *port;
+ * bool file_read = PolicyServerParseFile(GetWorkDir(), &host, &port);
+ * printf( "host is %s", file_read ? host : "unavailable" );
+ * free(host); free(port);
+ * @endcode
+ *
+ * @param[in]  workdir The directory of policy_server.dat usually GetWorkDir()
+ * @param[out] host pointer at this address will be hostname string (strdup)
+ * @param[out] port pointer at this address will be port string (strdup)
+ * @attention  host* and port* must be freed.
+ * @return     Contents of policy_server.dat file. Null terminated.
+ */
+bool PolicyServerParseFile(const char *workdir, char **host, char **port)
+{
+    char *contents = PolicyServerReadFile(workdir);
+    if (contents == NULL)
+    {
+        return false;
+    }
+    (*host) = NULL;
+    (*port) = NULL;
+
+    ParseHostPort(contents, host, port);
+    (*host) = xstrdup(*host);
+    if (*port != NULL)
+    {
+        (*port) = xstrdup(*port);
+    }
+    free(contents);
+    return true;
+}
+
+/**
+ * @brief      Reads and parses the policy_server.dat file.
+ *
+ * @param[in]  workdir The directory of policy_server.dat usually GetWorkDir()
+ * @param[out] ipaddr pointer at this address will be hostname string (strdup)
+ * @param[out] port pointer at this address will be port string (strdup)
+ * @attention  ipaddr* and port* must be freed.
+ * @return     Contents of policy_server.dat file. Null terminated.
+ * @see        PolicyServerParseFile
+ */
+bool PolicyServerLookUpFile(const char *workdir, char **ipaddr, char **port)
+{
+    char *host;
+    bool file_read = PolicyServerParseFile(workdir, &host, port);
+    if (file_read == false)
+    {
+        return false;
+    }
+    char tmp_ipaddr[CF_MAX_IP_LEN];
+    if (Hostname2IPString(tmp_ipaddr, host, sizeof(tmp_ipaddr)) == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Unable to resolve policy server host: %s", host);
+        return false;
```

Comment:
You need `free(host)` here as well, and I *think* you need `free(*port)` also. I assume that if you return false, the caller will not free it.